### PR TITLE
[Consensus Voting] Integration tests for building & verifying QC

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/dapperlabs/flow-go/issues/5970

### Context

This PR implements some improvements over https://github.com/onflow/flow-go/pull/1194 and adds integration tests for vote processors. Integrated tests were created to make sure that combined and staking vote processors produce a valid QC with real signatures and that QC can be verified by followers logic.

Notable changes:
- Replaced `Merger` with `Packer`.
- Added integration tests for combined & staking vote processors. 
- CombinedVoteProcessorV2 and CombinedVoteProcessorV3 tests are skipped since they are missing `RandomBeaconReconstructor` implementation.

Note about `RandomBeaconReconstructor`, implementation is ready at this point and there will be a following PR which completes the test and adds the implementation itself.

### CODE REVIEW

**Review by tarakby — APPROVED** [review]
Thanks for adding the tests 🏄🏼

File: consensus/hotstuff/votecollector/combined_vote_processor_v3_test.go [hunk 1]
```diff
@@ -752,3 +759,113 @@ func TestCombinedVoteProcessorV3_PropertyCreatingQCLiveness(testifyT *testing.T)
 		}
 	})
 }
+
+// TestCombinedVoteProcessorV2_BuildVerifyQC tests a complete path from creating votes to collecting votes and then
+// building & verifying QC.
+// We start with leader proposing a block, then new leader collects votes and builds a QC.
+// Need to verify that QC that was produced is valid and can be embedded in new proposal.
+func TestCombinedVoteProcessorV3_BuildVerifyQC(t *testing.T) {
+	t.Skip("This test is skipped due to missing RandomBeaconReconstructor impl")
```
Comment (durkmurder) [inline consensus/hotstuff/votecollector/combined_vote_processor_v3_test.go:768]: This test is majorly updated in https://github.com/onflow/flow-go/pull/1602, it's more of `StakingVoteProcessor` copy cat in this PR.

File: consensus/hotstuff/verification/staking_signer.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+// It returns:
+//  - (stakingSig, nil) signature signed with staking key.  The sig is 48 bytes long
+//  - (nil, error) if there is any exception
+func (c *StakingSigner) genSigData(block *model.Block) ([]byte, error) {
+	// create the message to be signed and generate signatures
+	msg := MakeVoteMessage(block.View, block.BlockID)
+
+	stakingSig, err := c.staking.Sign(msg, c.stakingHasher)
+	if err != nil {
+		return nil, fmt.Errorf("could not generate staking signature: %w", err)
```
Comment (zhangchiqing) [inline consensus/hotstuff/verification/staking_signer.go:93]:
```suggestion
		return nil, fmt.Errorf("could not generate staking signature for block (%v) at view %v: %w", block.BlockID, block.View, err)
```

File: consensus/hotstuff/verification/combined_verifier_v2.go [hunk 3]
```diff
@@ -1,3 +1,4 @@
+//go:build relic
```
Comment (zhangchiqing) [inline consensus/hotstuff/verification/combined_verifier_v2.go:1]: Is this needed because of the use of [crypto.NewBLSKMAC](https://github.com/onflow/flow-go/pull/1601/files#diff-0cd49adfc8c2df8b741ff1252f2dc39210c56f5f9a506d21154e259b01b66cdeR40)?

If so, we need to add it for [staking signer as well](https://github.com/onflow/flow-go/pull/1601/files#diff-d8863e5cd779af826a1f1a6d5118d62905a89d3a7bdb5eb8833aa35e004ac51eR34)

File: consensus/hotstuff/verification/staking_signer.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	return vote, nil
+}
+
+// genSigData generates the signature data for our local node for the given block.
+// It returns:
+//  - (stakingSig, nil) signature signed with staking key.  The sig is 48 bytes long
+//  - (nil, error) if there is any exception
+func (c *StakingSigner) genSigData(block *model.Block) ([]byte, error) {
```
Comment (zhangchiqing) [inline consensus/hotstuff/verification/staking_signer.go:87]: I think we should be able to merge this module with [CombinedSigner](https://github.com/onflow/flow-go/blob/leo/sign-vote/consensus/hotstuff/verification/combined_signer.go).

Since the only difference between them is the `genSigData`, which can be passed in as dependency.

If you also agree, we can do it in a separate PR. Doesn't have to be done in this PR

**Review by zhangchiqing — APPROVED** [review]
Looks good 👍

File: consensus/hotstuff/verification/combined_verifier_v2.go (hunk 3 above)
Comment (durkmurder) [reply consensus/hotstuff/verification/combined_verifier_v2.go:1]: I think GoLand adds it automatically

File: consensus/hotstuff/verification/staking_signer.go (hunk 4 above)
Comment (durkmurder) [reply consensus/hotstuff/verification/staking_signer.go:87]: but the `CombinedSigner` gets removed?

**Comment by durkmurder** [comment]
bors merge

**Comment by durkmurder** [comment]
bors retry
